### PR TITLE
[mlir] Introduce `trailingNDimsContiguous` for MemRefs

### PR DESCRIPTION
**Pull request by banach-space**
Extracts logic from `vector::isContiguousSlice` to check whether
the trailing dim of a memref are contiguous into a dedicated hook
in BuiitinTypes.{h|cpp}.

Follow-up for https://github.com/llvm/llvm-project/pull/76848.

### CODE REVIEW

File: mlir/lib/IR/BuiltinTypes.cpp [hunk 1]
```diff
@@ -967,3 +967,32 @@ bool mlir::isLastMemrefDimUnitStride(MemRefType type) {
...
 }
+
+bool mlir::trailingNDimsContiguous(MemRefType type, int64_t n) {
+  if (!isLastMemrefDimUnitStride(type))
+    return false;
+
+  auto memrefShape = type.getShape().take_back(n);
+  int64_t offset;
+  SmallVector<int64_t> stridesFull;
+  if (!succeeded(getStridesAndOffset(type, stridesFull, offset)))
```
Comment (dcaballe) [inline mlir/lib/IR/BuiltinTypes.cpp:978]: Is the memref layout being taken into account anywhere? That's one of the differences between vector and memref.

Comment (banach-space) [reply mlir/lib/IR/BuiltinTypes.cpp:984]: I've just added a check that leverages [isIdentity](https://github.com/llvm/llvm-project/blob/e3172e841899779caf461cee97826940dca806e7/mlir/include/mlir/IR/BuiltinAttributeInterfaces.td#L475-L482). 

> Is the memref layout being taken into account anywhere?

I thought that that's what `getStridesAndOffset` does (i.e. returns the layout as offsets and strides)? I am not 100% sure though, this is mostly just moving things around.

**Comment by banach-space** [comment]
Kind ping :)

File: mlir/include/mlir/IR/BuiltinTypes.h [hunk 2]
```diff
@@ -518,6 +518,16 @@ bool isStrided(MemRefType t);
...
 
+/// Return "true" if the last N dimensions of the given type are contiguous.
+///
+/// Examples:
+///   - memref<5x4x3x2xi8, strided<[24, 6, 2, 1]> is contiguous when
+///   considering both _all_ and _only_ the trailing 3 dims,
+///   - memref<5x4x3x2xi8, strided<[48, 6, 2, 1]> is _only_ contiguous when
+///   considering the trailing 3 dims.
+///
+bool trailingNDimsContiguous(MemRefType type, int64_t n);
```
Comment (dcaballe) [inline mlir/include/mlir/IR/BuiltinTypes.h:529]: I couldn't infer by the name what this was doing. Could we rename this to something like `isContiguousSlice` or something starting with `isContiguous...`?

**Review by dcaballe — APPROVED** [review]
LGTM! Just a minor comment

**Pull request merged**